### ISSUE TITLE
CUDA Out of Memory during 3d_fullres inference on RTX 5090 (32 GB VRAM, Windows/WDDM)

### ISSUE THREAD

**Issue by jdmerkur** [MIC-DKFZ/nnUNet#2963]
Firstly just wanted to say thank you for all the work put into this segmentation model, and making it publicly accessible. Really appreciate all of the work that has gone into nnU-Net

Description

We encounter reproducible CUDA out-of-memory (OOM) errors during 3d_fullres inference in nnU-Net v2 despite using an RTX 5090 with ~32 GB VRAM. Inference is run conservatively (single fold, no TTA, minimal workers).

One important characteristic of our task is a large number of segmentation labels, which may significantly increase memory usage during softmax, argmax, and resampling/export of multi-class probability maps. The same dataset runs successfully with 3d_lowres.

The OOM typically occurs after several cases complete, suggesting memory accumulation or incomplete VRAM release between cases, rather than a single-volume forward-pass failure.

Environment:

OS: Windows (WDDM)
GPU: NVIDIA GeForce RTX 5090
VRAM: 32607 MiB (~32 GB)
Driver: 576.83
CUDA (driver): 12.9
CUDA (PyTorch): 12.8
Python: 3.13.10 (Anaconda)
PyTorch: 2.8.0+cu128
nnU-Net: nnUNetv2 (latest)

This environment uses Python 3.13.10, which is newer than typical nnU-Net setups. If this is a known limitation, we are happy to test on Python 3.10/3.11 for comparison; however, the observed behavior suggests a memory accumulation issue rather than a single-allocation peak.

**Comment by mrokuss** [comment]
Hey @jdmerkur and @woojinchokimm 

Your assumption is likely correct: the combination of many output classes and a large 3D patch size significantly increases memory consumption during prediction. One thing you can try: switch to a more memory-efficient torch-based resampling method instead of the default skimage implementation.

### Memory Consumption Causes
During prediction, memory usage scales dramatically with the image size and the number of classes because the softmax calculation and result aggregation create massive probability maps of size `N_classes x X x Y x Z`. nnUNet strongly favors large patch sizes, which dominates the VRAM usage during these multi-class aggregation operations. Furthermore, the default inference pipeline uses `skimage` for resampling, which is known to be highly memory-intensive and can cause backlogs that lead to out-of-memory errors.

### Torch Resampling Fix
You can enable an alternative resampling function that uses `torch.interpolate` to significantly reduce RAM and VRAM usage at the expense of a minuscule reduction in segmentation accuracy. To apply this quick fix, you must modify the configurations in your plans files located in both your preprocessed folder and results folder. Replace the default probabilities resampling block with the torch-based configuration below, which will take effect immediately.

**Replace:**
```json
"resampling_fn_probabilities": "resample_data_or_seg_to_shape",
"resampling_fn_probabilities_kwargs": {
    "is_seg": false,
    "order": 1,
    "order_z": 0,
    "force_separate_z": null
}
```

**With:**
```json
"resampling_fn_probabilities": "resample_torch_fornnunet",
"resampling_fn_probabilities_kwargs": {
    "is_seg": false,
    "force_separate_z": false,
    "memefficient_seg_resampling": false
}
```

Hope this helps!

**Comment by 1yuechen-github1** [comment]
> 首先，想感谢大家为这个细分模型所做的所有努力，以及让它对公众开放。非常感谢为nnU-Net付出的所有努力
> 
> 描述
> 
> 尽管使用RTX 5090，显存~32GB，我们在nnU-Net v2的推理中仍遇到可复现CUDA内存外（OOM）错误3d_fullres。推理采用保守方式（单重折叠，无TTA，最小工作者）。
> 
> 我们任务的一个重要特点是大量分割标签，这在软最大、argmax以及多类概率图的重采样/导出过程中，可能显著增加内存使用。同一个数据集也能成功运行3d_lowres。
> 
> OOM通常发生在多个案例完成后，表明内存积累或案例间的显存释放不完整，而非单卷前向传递失败。
> 
> 环境：
> 
> 操作系统：Windows（WDDM） GPU：NVIDIA GeForce RTX 5090 显存：32607 MiB（~32 GB） 驱动：576.83 CUDA（驱动）：12.9 CUDA（PyTorch）：12.8 Python：3.13.10（Anaconda） PyTorch：2.8.0+cu128 nnU-Net：nnUNetv2（最新）
> 
> 该环境使用 Python 3.13.10，比典型的 nnU-Net 配置更新。如果这是已知的限制，我们愿意在 Python 3.10/3.11 上进行测试以作对比;然而，观察到的行为表明是记忆积累问题，而非单一分配峰值。

我发现是不同ct的spacing不同导致部分ct比较大，试了把spacing全部同一到0.3就解决了

**Comment by woojinchokimm** [comment]
Thanks @mrokuss, sadly I did not try your solution, but it does make sense. 

While debugging I realized that if your original image size and number of classes is large, the resampling step causes OOM. In my case I was working with CBCT volumes of size (300,300,300) with 31 classes!

I don't have a code snippet at hand, but what helped me was changing the order of the resampling.

Original nnunetv2 order:
logits (C, X, Y, Z) --> resample logits --> argmax --> segmentation (X, Y, Z)

New order:
logits (C, X, Y, Z) --> argmax --> segmentation (X, Y, Z) --> resample segmentation

Instead of resampling the full logits tensor, I computed the segmentation (argmax) first and only then resampled the resulting label map. My peak memory usage went down to 4-6GB during the resampling step. (If running with docker you might have to allow increased memory).

Hope that's clear @jdmerkur
